### PR TITLE
enable multiple steamaccounts and librarypaths

### PR DESCRIPTION
**Pull request by rokam**

### CODE REVIEW

**Comment by geeksville** [comment]
That looks like some really great changes.  Can you add some notes here on goals and what the changes were?  

(And I super appreciate the PR but changes to formatting (i.e. string delimiters etc) makes it hard to review / see where there are new improvements / risk of bugs)

File: defaults/py_modules/steamback/__init__.py [hunk 1]
```diff
@@ -31,6 +31,18 @@ def _parse_vcf(path: str) -> dict:
     return d
 
 
+def _parse_libs(path: str) -> list[str]:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:34]: Method to get library paths from steamapps/libraryfolders.vdf

File: defaults/py_modules/steamback/__init__.py [hunk 2]
```diff
@@ -50,35 +62,39 @@ def __init__(self, config: Config):
         #    logger.info('Running under decky')
         # except:
 
-        logger.info(f'Steamback engine created: { config }')
+        logger.info(f"Steamback engine created: { config }")
 
         # a dict from gameid -> gameinfo for all installed games.  ONLY USED ON DESKTOP not DECKY
         self.all_games = None
-        self.account_id = 0
+        self.account_ids: list[int] = []
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:69]: Multiple account ids

File: defaults/py_modules/steamback/__init__.py [hunk 3]
```diff
@@ -50,35 +62,39 @@ def __init__(self, config: Config):
...
         self.dry_run = False  # Set to true to suppress 'real' writes to directories
         self.max_saves = 10  # default to a max of ten saves
 
         # don't generate backups if the files haven't changed since last backup
         self.ignore_unchanged = True
 
-    def set_account_id(self, id_num: int):
-        logger.debug(f'Setting account id { id_num } on { self }')
-        self.account_id = id_num
+    def add_account_id(self, id_num: int):
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:76]: Change set account to add accoount

File: defaults/py_modules/steamback/__init__.py [hunk 4]
```diff
@@ -50,35 +62,39 @@ def __init__(self, config: Config):
...
+        self.account_ids.append(id_num)
 
     """Find the steam account ID for the current user (and)
 
     We look for a directory in userdata/accountname.  If there are multiple account names we currently throw an exception
     eventually we could look at the config file in each of those dirs to see which one was touched most recently.
     """
 
-    def auto_set_account_id(self) -> int:
+    def auto_set_account_id(self) -> list[int]:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:86]: Populate the list of accounts

File: defaults/py_modules/steamback/__init__.py [hunk 5]
```diff
@@ -103,9 +119,14 @@ def get_steam_root(self) -> str:
     Return the path to the game directory for a specified game
     """
 
-    def _get_gamedir(self, game_id: int) -> str:
-        assert self.account_id  # better be set by now
-        return os.path.join(self.get_steam_root(), "userdata", str(self.account_id), str(game_id))
+    def _get_gamedir(self, game_id: int) -> list[str]:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:122]: Return the possible game paths, based on each account

File: defaults/py_modules/steamback/__init__.py [hunk 6]
```diff
@@ -130,35 +151,44 @@ def _get_steamapps_dir(self, game_info: dict, is_system_dir: bool = False) -> st
...
         app_dir = self._get_steamapps_dir(game_info)
-        vcf = _parse_vcf(os.path.join(
-            app_dir, f'appmanifest_{ game_info["game_id"] }.acf'))
+        vcf = _parse_vcf(
+            os.path.join(app_dir, f'appmanifest_{ game_info["game_id"] }.acf')
+        )
         install_dir = vcf.get("installdir", None)
         return install_dir
 
+    def _get_all_library(self) -> list[str]:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:160]: Method to retrive all libraries

File: defaults/py_modules/steamback/__init__.py [hunk 7]
```diff
@@ -130,35 +151,44 @@ def _get_steamapps_dir(self, game_info: dict, is_system_dir: bool = False) -> st
...
-                id = int(id)
-                info = {
-                    # On a real steamdeck there may be multiple install_roots (main vs sdcard etc) (but only one per game)
-                    "install_root": steam_dir,
-                    "game_id": id,
-                    "game_name": name
-                }
-                r.append(info)
-                rdict[id] = info
+        for steam_dir in self._get_all_library():
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:171]: Same logic using each library to find the game manifest

File: defaults/py_modules/steamback/__init__.py [hunk 8]
```diff
@@ -281,20 +328,24 @@ def addRoots(is_system_dir: bool):
...
-            # Store the savegames directory for this game
-            return remoteFound
-
-        # okay - now check the standard doc roots for games - do this before looking for autoclouds because it is more often the match
-        d = self._search_likely_locations(game_info, rcf)
-        if (d):
-            return d
+        dirs = self._get_gamedir(game_info["game_id"])
+
+        for d in dirs:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:333]: Loop through possible userdata folders, one  for each account

File: defaults/py_modules/steamback/__init__.py [hunk 9]
```diff
@@ -334,60 +385,58 @@ def _rcf_is_valid(self, root_dir: str, rcf: list[str]):
...
-                            prevl = None
-
-                        # Now skip until we get a close brace
-                        skipping = True
-                    else:
-                        prevl = s
-        else:
-            logger.debug(f'No rcf {path}')
-            return None
+        for path in paths:
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:393]: As we now have multiple userdata paths, it loops through all of them looking for the remotecache.vdf file.

Comment (rokam) [reply defaults/py_modules/steamback/__init__.py:393]: All the logic is still the same

File: defaults/py_modules/steamback/__init__.py [hunk 10]
```diff
@@ -334,60 +385,58 @@ def _rcf_is_valid(self, root_dir: str, rcf: list[str]):
...
+                                filename = (prevl[1:])[:-1]
+                                rcf.append(filename)
+                                prevl = None
+
+                            # Now skip until we get a close brace
+                            skipping = True
+                        else:
+                            prevl = s
+            else:
+                logger.debug(f"No rcf {path}")
```
Comment (rokam) [inline defaults/py_modules/steamback/__init__.py:423]: Removed the return  None. as now it need to search  inside the for

Comment (rokam) [reply defaults/py_modules/steamback/__init__.py:423]: And the rcf will remain empty, so the rest of the logic won't find any files to backup

File: defaults/py_modules/steamback/test.py [hunk 11]
```diff
@@ -9,7 +9,7 @@
 async def testImpl(p: Engine):
     print('Simulating decky loader for testing...')
 
-    p.set_account_id(49847735)
+    p.add_account_id(49847735)
```
Comment (rokam) [inline defaults/py_modules/steamback/test.py:12]: renamed the method

**Review by rokam — COMMENTED** [review]
Here I comment on every change. I'm sorry for formating the code.

**Comment by geeksville** [comment]
ok great thanks!  I'll work on this tonight.  The fix sounds great but I was scared because omg the diffs were so huge.

**Comment by geeksville** [comment]
looks a lot better now!

**Comment by geeksville** [comment]
looks great!  thanks for the PR I'm sure people will love it.

**Pull request merged**